integrationtest/integration_test.go

```diff
@@ -18,26 +20,56 @@ import (
 //   - A Kubernetes cluster in which:
 //       - it can use the `test` namespace
 //       - Vault is deployed and accessible
+//       - There is a serviceaccount called test-token-reviewer-account with access to the TokenReview API
 // See `make setup-integration-test` for manual testing.
 func TestMain(m *testing.M) {
 	if os.Getenv("INTEGRATION_TESTS") != "" {
-		cmd := exec.Command("kubectl", "exec", "--namespace=test", "vault-0", "--", "cat", "/var/run/secrets/kubernetes.io/serviceaccount/token")
-		out := &bytes.Buffer{}
-		cmd.Stdout = out
-		cmd.Stderr = out
-		if err := cmd.Run(); err != nil {
-			fmt.Println(out.String())
-			fmt.Println(err)
-			os.Exit(1)
-		}
 		os.Setenv("VAULT_ADDR", "http://127.0.0.1:38200")
 		os.Setenv("VAULT_TOKEN", "root")
-		os.Setenv("KUBERNETES_JWT", out.String())
+		os.Setenv("KUBERNETES_JWT", getVaultServiceAccountJWT())
+		os.Setenv("TOKEN_REVIEWER_JWT", getTokenReviewerJWT())
 		os.Exit(m.Run())
 	}
 }
 
-func TestUnauthorizedServiceAccountErrorCode(t *testing.T) {
+// TODO: In 1.24 this will break because k8s will stop auto-generating tokens for service accounts:
+// https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2799-reduction-of-secret-based-service-account-token#proposal.
+// The cleanest long-term solution will probably be to use TokenRequest to generate our own tokens on demand.
+// Unfortunately that would take a little more boiler plate than usual because kubectl doesn't support it directly and
+// we don't want to import k8s.io/client-go, as its dependencies have caused issues in the past in upstream Vault.
+// We could do kubectl config --raw -o json and parse the result into a struct, then extract the
+// API/CA/auth information required from there to interact with the k8s API directly. There's a good chance there's a
+// cleaner/simpler way too.
+func getTokenReviewerJWT() string {
+	name := runCmd("kubectl --namespace=test get serviceaccount test-token-reviewer-account -o jsonpath={.secrets[0].name}")
+	b64token := runCmd(fmt.Sprintf("kubectl --namespace=test get secrets %s -o jsonpath={.data.token}", name))
+	token, err := base64.URLEncoding.DecodeString(b64token)
+	if err != nil {
+		panic(err)
+	}
+	return string(token)
+}
+
+func getVaultServiceAccountJWT() string {
+	return runCmd("kubectl exec --namespace=test vault-0 -- cat /var/run/secrets/kubernetes.io/serviceaccount/token")
+}
+
+func runCmd(command string) string {
+	parts := strings.Split(command, " ")
+	cmd := exec.Command(parts[0], parts[1:]...)
+	out := &bytes.Buffer{}
+	cmd.Stdout = out
+	cmd.Stderr = out
+	if err := cmd.Run(); err != nil {
+		panic(fmt.Sprintf("Got unexpected output: %s, err = %s", out.String(), err))
+	}
+	return out.String()
+}
+
+func setupKubernetesAuth(t *testing.T, boundServiceAccountName string,
+	kubeConfigOverride map[string]interface{},
+) (func(), *api.Client) {
```

Comment:
nit: The line breaks here look weird for Go style to me: https://github.com/golang/go/wiki/CodeReviewComments#line-length

Reply:
It's weird, but it is what `gofumpt -w .` gave me. I put it all in one line and it seems to allow that. Otherwise it puts the return signature on the end like this. 🤷

Reply:
Huh, TIL